src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -33,55 +31,84 @@ internal abstract partial class AbstractNavigateToSearchService
         private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
             Project project, Document searchDocument, string pattern, CancellationToken cancellationToken)
         {
-            var containsDots = pattern.IndexOf('.') >= 0;
-            using (var patternMatcher = new PatternMatcher(pattern, allowFuzzyMatching: true))
+            // If the user created a dotted pattern then we'll grab the last part of the name
+            var (patternName, patternContainerOpt) = PatternMatcher.GetNameAndContainer(pattern);
+            var nameMatcher = PatternMatcher.CreatePatternMatcher(patternName, includeMatchedSpans: true, allowFuzzyMatching: true);
+
+            var containerMatcher = patternContainerOpt != null
+                ? PatternMatcher.CreateDotSeperatedContainerMatcher(patternContainerOpt)
+                : null;
+
+            using (nameMatcher)
+            using (containerMatcher)
             {
-                var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
-                foreach (var document in project.Documents)
+                var nameMatches = ArrayBuilder<PatternMatch>.GetInstance();
+                var containerMatches = ArrayBuilder<PatternMatch>.GetInstance();
+
+                try
                 {
-                    if (searchDocument != null && document != searchDocument)
-                    {
-                        continue;
-                    }
+                    return await FindNavigableDeclaredSymbolInfosAsync(
+                        project, searchDocument, nameMatcher, containerMatcher,
+                        nameMatches, containerMatches, cancellationToken).ConfigureAwait(false);
+                }
+                finally
+                {
+                    nameMatches.Free();
+                    containerMatches.Free();
+                }
+            }
+        }
+
+        private static async Task<ImmutableArray<INavigateToSearchResult>> FindNavigableDeclaredSymbolInfosAsync(
+            Project project, Document searchDocument,
+            PatternMatcher nameMatcher, PatternMatcher containerMatcher,
+            ArrayBuilder<PatternMatch> nameMatches, ArrayBuilder<PatternMatch> containerMatches,
+            CancellationToken cancellationToken)
+        {
+            var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+            foreach (var document in project.Documents)
+            {
+                if (searchDocument != null && document != searchDocument)
+                {
+                    continue;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
i have measured the differences with/without hte checks and found no actual measurable difference in our scenarios.  I prefer to not have to spend inordinate amount of time thinking about things that don't end up actually mattering in practice.  As such, following simple patterns means that code is consistent and well behaved, without being microoptimized with the need to then have to think if that is/isn't ok in the future.